### ISSUE TITLE
bar_width broken for Date/DateTime x axis

### ISSUE THREAD

**Issue by mkborregaard** [JuliaPlots/Plots.jl#748]
When the x axis is DateTime, it is no longer possible to set `bar_width`. The default value (`nothing`) works, but it is not possible to specify it. The reason seems to be that the x axis becomes reinterpreted to become very large, corresponding to the Int value of the number as DateTime. It would be nice with some more useful defaults.

```julia
bar(Date.(1:5), 1:5) #works
bar(Date.(1:5), 1:5, bar_width = 1) # does not work
bar(Date.(1:5), 1:5, bar_width = 1e10) # sort of works

julia> Int.(Date.(1:5))
5-element Array{Int64,1}:
    1
  366
  731
 1096
 1462
julia> Int.(DateTime.(1:5))
5-element Array{Int64,1}:
     86400000
  31622400000
  63158400000
  94694400000
 126316800000

julia> Int.(DateTime.(1:5)) / 1e10
5-element Array{Float64,1}:
  0.00864
  3.16224
  6.31584
  9.46944
 12.6317 
```
@daschw I think this is caused by https://github.com/JuliaPlots/Plots.jl/pull/720

**Comment by daschw** [comment]
OK, I see. #720 definitely changed the `bar_with` behaviour for `Date` and I can work on reversing this. 

However, I think the issue existed before: I changed the plot recipes for `Date` and `DateTime` on my local Plots install back to
```
@recipe f(::Type{Date}, dt::Date) = (dt -> convert(Int,dt), dt -> string(convert(Date,dt)))
@recipe f(::Type{DateTime}, dt::DateTime) = (dt -> convert(Int,dt), dt -> string(convert(DateTime,dt)))
```
as they were before #720 and
```
bar(Date.(1:5), 1:5, bar_width = 1)
```
did not work as one would expect either. With `bar_width = 1` you would expect the bars to be directlly next to each other without any space between, right?
I could achieve this with
```
bar(Date.(1:5), 1:5, bar_width = 365)
```
so I suppose that a `bar_width` of 1 corresponded to 1 day on the x axis. Now, after #720 it corresponds to 1 millisecond, so you would achieve the same with
```
bar(Date.(1:5), 1:5, bar_width = 365 * 864e5)
```
Anyway, I may be mistaken, so please correct me, if you are sure that this was different before #720!

**Comment by mkborregaard** [comment]
No I think you are right on the money. For integer `Date` objects, the distance between bars on the x axis is in years and the unit of bar_width used to be days, today they are milliseconds. While the former might be more generally useful than the latter, none of them are ideal. So as you say, it is more precise to say that the behaviour was changed with #720 than that this PR is the root of all the trouble :-)

Anyway, it'd be great if you would look at the behaviour here - at present the change has broken all plots recipes using bars and times (which may only be the one I added to TimeSeries :-) but who knows what is out there in the wild?).
Perhaps it is hard to establish the most useful unit of the '1' bar_width for TimeTypes, where the scales are nested, though? (I mean doesn't `Int.(Date.(4)) == 1096` say it all? Converting from Int to Date back to Int gives completely noncongruent results :-) ). It seems to me that the most useful unit would be the unit of the integer on the X axis, perhaps?

**Comment by daschw** [comment]
OK, I will first work on restoring the old behaviour for `Date` objects.

> I mean doesn't Int.(Date.(4)) == 1096 say it all?

`Int(convert(Date, 4)) == 4`. `Date(4)` constructs the `Date` object `0004-01-01` and does not convert the Integer 4 to its `Date` equivalent. In the recipes for `Date` and `DateTime` `convert()` is used so the noncongruent results, you mentioned, should not really be related to the `bar_width` issue.

> none of them are ideal

I agree, but somehow it is the expected behaviour, right? I assumed before that a `bar_width = 1` would be expected to be equivalent to `bar_width = nothing`, but I guess it really is intended to match the unit length on the x axis.

Would the introduction of a `relative_bar_width` attribute between 0 and 1 make sense for this type of scales?

**Comment by mkborregaard** [comment]
Yes I realize it calls the constructor, not convert, I just find it confusing. WRT `bar_width = nothing` I think that needs editing. Bars are not supposed to touch by default (but they do now) - in fact, bars in bar diagrams never touch in practice. Otherwise they are too easy to confuse with histograms.

I'd rather not introduce a `relative_bar_width`. I think it shoud ideally be in the units on the x axis throughout. That is tricky with times, but I think at least a unit of days is more generally useful than milliseconds for Date. Perhaps the processing of `nothing` could do something sensible based on 0.8 * minimum(diff(x)) or something like that?

**Comment by daschw** [comment]
Oh, I should have clicked on the link you provided before :-) I see that you already knew `Int(convert(Date, 4)) == 4`

Days for `Date` definitely make sense, but what would be the default unit for `DateTime`? Seconds? Hours? Days? For consistency it still should be Millisecond imho.

**Comment by mkborregaard** [comment]
The alternative is to make bar_width always relative - not sure what I prefer in fact?

**Comment by daschw** [comment]
Well, this would probably break a lot of existing plots ...

**Comment by mkborregaard** [comment]
good point

**Comment by mkborregaard** [comment]
Closed by #752
